스프링 7주차 과제 : 인증, 암호화

bloomspes opened the pull request:
[이번 주 목표]
- Spring Security 프레임워크를 사용하여 인증, 암호화 구현하기.
- 2~6주차 과제 반복학습

bloomspes committed:
build : installed spring security

bloomspes committed:
add : Security configuration 코드를 작성하라.

- 평문 통신을 위해 csrf 공격 disable 설정 추가
- spring security filter 추가.

bloomspes committed:
add : spring security - authenticationManager 작성

- security configuration > authenticationManager 인자 받기
- jwt Auth filter 에서는 basic Auth filter 상속 받아서 super 메소드로 전달.

bloomspes committed:
add : spring security - doFilterInternal()

http request, response를 받아서 연쇄적으로 인증단계를 진행하기.

bloomspes committed:
refactor : interceptors -> jwtAuthenticationFilter 인증 분리

bloomspes committed:
refactor : authentication invalid Token exception 로직을 작성하라.

bloomspes committed:
refactor : filterWithPathAndMethod 를 실행하고 chain 메소드를 실행하도록 수정

hannut91 commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,25 @@
+        Filter authenticationFilter = new JwtAuthenticationFilter(
+                authenticationManager(), authenticationService);
+        http
+                .csrf().disable()
```
CSRF는 무엇이고 여기서 왜 disable하는 것인가요?

bloomspes replied:
CSRF : 사이트 요청을 위조하는 공격입니다. `POST`, `PUT`, `MATCH` 메소드에 해당하는 패킷을 가로채서 해당하는 요청을 바꾸는 공격인데, 강의에서도 과제 수행시에 평문 통신을 구현하고 있어서 csrf 보안에 사용되는 토큰을 disable 메소드로 off 시키는 것으로 현재 이해했습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,67 @@
+            authenticationService.parseToken(accessToken);
+        } catch (InvalidTokenException e) {
+            response.sendError(HttpStatus.UNAUTHORIZED.value());
+            return;
```
여기에는 return이 필요하지 않아 보이네요

bloomspes replied:
반영하겠습니다 :)

bloomspes committed:
refactor : interceptors 삭제, WebJavaConfig 클래스 삭제

hannut91 commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -1,25 +1,32 @@
 @Configuration
 public class SecurityJavaConfig extends WebSecurityConfigurerAdapter {
-    @Autowired
-    private AuthenticationService authenticationService;
```
Autowired는 왜 삭제하셨나요? 삭제해도 아무런 영향이 없나요? 없다면 왜 없을까요?

bloomspes replied:
**Q1. Autowired는 왜 삭제 하셨나요?**

@hannut91 IDE에서 Autowired를 권장하지 않는다고 해서 리팩토링을 실행했더니 생성자 필드를 만들어주더라구요, 여기에 대해서 안그래도 종립님께 질문을 드렸는데, 두 가지 이유를 말씀해 주셨어요.
1. 스프링 개발팀이 생성자 방식을 권고한다.
2. Autowired는 reflection을 사용해서 객체에 의존 관계를 주입하기 때문에 이런 특성으로 의존성 주입에 문제가 생길 수 있는 여지가 존재하는 점 이였습니다.

저도 잘 모르고 있었던 건데, 이번 기회에 Autowired와 자바 어노테이션에 대해서 공부를 더 하게 되었어요.

**Q2. 삭제해도 아무런 영향이 없을까요? 없다면 왜 없을까요?**

Autowired 대신에 생성자 필드를 선언했기 때문에, 생성자를 통한 객체의 의존 주입이 가능해집니다.
따라서 Autowired의 기능을 대체할 수 있습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -16,8 +15,8 @@
-            AuthenticationManager authenticationManager,
-            AuthenticationService authenticationService) {
+            AuthenticationService authenticationService,
+            AuthenticationManager authenticationManager) {
```
왜 순서를 변경하셨나요? 보면 authenticationManager가 먼저 쓰이고 있어서요

bloomspes replied:
수정했습니다. :)

hannut91 commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -54,14 +50,7 @@ private boolean filterWithPathAndMethod(HttpServletRequest request) {
-
-        if (method.equals("OPTIONS")) {
-            return true;
-        }
```
이것은 왜 삭제하셨나요?

bloomspes replied:
아, 이부분은 리팩토링 하다가 제 실수로 지워졌습니다...

bloomspes committed:
refactor : 코드리뷰 반영

hannut91 commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        Filter authenticationFilter = new JwtAuthenticationFilter(
+                authenticationManager(), );
```
```suggestion
        Filter authenticationFilter = new JwtAuthenticationFilter(
                authenticationManager());
```

여기 콤마가 잘못들어갔어요

hannut91 commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,59 @@
+
+        if (authorization == null) {
+            throw new InvalidTokenException("");
+        }
```
관련이 있는 코드끼리 잘 묶어주세요
```suggestion
        String authorization = request.getHeader("Authorization");
        if (authorization == null) {
            throw new InvalidTokenException("");
        }
```

bloomspes committed:
Update app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java

Co-authored-by: Yunseok <hannut91@gmail.com>

bloomspes committed:
Update app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFilter.java

Co-authored-by: Yunseok <hannut91@gmail.com>

bloomspes committed:
refactor : 코드리뷰 반영

bloomspes committed:
Merge branch 'main' of https://github.com/bloomspes/spring-week7-assignment-1

bloomspes committed:
add : spring security 암호화 강의 학습

password 암호화 하기

bloomspes committed:
add : password 인증 암호화 - 테스트코드 작성

bloomspes committed:
add : user domain에 password encoder 객체로부터 암호화된 패스워드를 받는다.

bloomspes committed:
refactor : passwordEncoder 객체를 주입 받아서 서비스 클래스에서 사용하도록 변경

bloomspes committed:
refactor : UserService 객체에 passwordEncoder 객체 의존성을 주입하라.

bloomspes committed:
debugging : 테스트 코드 error 픽스중

bloomspes committed:
debug : authenticationService 파라미터 추가

bloomspes committed:
debug : error fixed.

bloomspes committed:
debug : ProductController 파라미터 수정, filter의 userid 처리 방식 수정

 Please enter the commit message for your changes. Lines starting

bloomspes committed:
과제 : 내 아이디로 인증되었을 때 내 정보를 수정할 수 있도록 변경하기

hannut91 commented:
CI실행된 결과를 보면 스프링 테스트에서 실패하고 있는 것 같습니다 ㅎㅎ 테스트를 한 번 확인해 보세요

bloomspes committed:
과제 : UserController, ProductController 테스트코드 디버깅

bloomspes committed:
과제: User 테스트코드 누락된 메소드 추가

bloomspes committed:
과제: ProductController 코드 수정

bloomspes committed:
과제 : UserService 코드 수정

bloomspes committed:
과제 : 스프링 테스트코드 수정

bloomspes committed:
과제 : 테스트 통과

bloomspes committed:
과제 : 테스트코드 모두 통과

bloomspes committed:
과제 : UserAuthenticated 테스트 커버리지 통과

bloomspes committed:
과제 : 커버리지 조정 - builder 무시

bloomspes committed:
Use JDK 16

Upgrade Gradle plugins:
- Spring Boot plugin
- Spring Dependency Management plugin

Delete JUnit 5 implementation statement, because it's in
`spring-boot-starter-test`.

Upgrade Gradle version from 6 to 7.
